--- a/lib/interface.c
+++ b/lib/interface.c
@@ -907,7 +907,7 @@
	short_rx = rx * 10;  
	short_tx = tx * 10;
	if (rx > 1152921504606846976ull) {
-	    short_rx /= 1152921504606846976ull;
+	    short_rx = rx / 115292150460684697ull;
	    Rext = "EiB";
	} else if (rx > 1125899906842624ull) {
	    short_rx /= 1125899906842624ull;
@@ -926,7 +926,7 @@
	    Rext = "KiB";
	}
	if (tx > 1152921504606846976ull) {
-	    short_tx /= 1152921504606846976ull;
+	    short_tx = tx / 115292150460684697ull;
	    Text = "EiB";
	} else if (tx > 1125899906842624ull) {
	    short_tx /= 1125899906842624ull;
