diff --git a/hostname.c.old b/hostname.c
index 742cfae..86e3ad2 100644
--- a/hostname.c.old
+++ b/hostname.c
@@ -44,6 +44,8 @@
 #include "config.h"
 #include "version.h"
 #include "../intl.h"
+#include "sys/types.h"
+#include "sys/socket.h"
 
 #if HAVE_AFDECnet
 #include <netdnet/dn.h>
@@ -128,6 +130,10 @@ static void setdname(char *dname)
 
 static void showhname(char *hname, int c)
 {
+    struct addrinfo *res;
+    struct addrinfo hints;
+    int ret;
+
     struct hostent *hp;
     register char *p, **alias;
     struct in_addr **ip;
@@ -135,13 +141,24 @@ static void showhname(char *hname, int c)
   if (c != 'I' && c != 'A') { // 'I' and 'A' don't depend on name resolution
     if (opt_v)
 	fprintf(stderr, _("Resolving `%s' ...\n"), hname);
-    if (!(hp = gethostbyname(hname))) {
+
+    memset(&hints, 0, sizeof(struct addrinfo));
+    hints.ai_socktype = SOCK_DGRAM;
+    hints.ai_flags = AI_CANONNAME;
+
+    if ((ret = getaddrinfo(hname, NULL, &hints, &res)) != 0) {
 	herror(program_name);
 	exit(1);
     }
+	
+       if(!(hp = gethostbyname(hname))) {
+          herror(program_name);
+          exit(1);
+       }
+
     if (opt_v) {
 	fprintf(stderr, _("Result: h_name=`%s'\n"),
-		hp->h_name);
+		res->ai_canonname);
 
 	alias = hp->h_aliases;
 	while (alias[0])
@@ -153,9 +170,10 @@ static void showhname(char *hname, int c)
 	    fprintf(stderr, _("Result: h_addr_list=`%s'\n"),
 		    inet_ntoa(**ip++));
     }
-    if (!(p = strchr(hp->h_name, '.')) && (c == 'd'))
+    if (!(p = strchr(res->ai_canonname, '.')) && (c == 'd'))
 	return;
   }
+
     switch (c) {
     case 'I':
     case 'A': {
@@ -231,7 +249,7 @@ static void showhname(char *hname, int c)
 	printf("%s\n", ++p);
 	break;
     case 'f':
-	printf("%s\n", hp->h_name);
+	printf("%s\n", res->ai_canonname);
 	break;
     default:
     ;
